File: src/main/java/endytkn/randomEvents/randomEvent/RandomEvent.java
package endytkn.randomEvents.randomEvent;

import endytkn.randomEvents.RandomEventsMod;
import endytkn.randomEvents.eventObservers.MinecraftEventsObservers;
import endytkn.randomEvents.utils.Observer;
import net.minecraft.core.BlockPos;
import net.minecraft.nbt.CompoundTag;
import net.minecraft.network.chat.Component;
import net.minecraft.server.level.ServerLevel;
import net.minecraft.server.level.ServerPlayer;
import net.minecraft.world.entity.Entity;
import net.minecraft.world.entity.player.Player;
import net.minecraftforge.event.TickEvent;
import net.minecraft.client.Minecraft;

import java.util.List;
import java.util.UUID;
import java.util.function.Consumer;

public class RandomEvent {
    public enum RandomEventStatus {
        NOT_STARTED, PREPARING, READY, WAITING_PLAYER, FINISHING_SUCCESS, FINISHING_CANCELED, FINISHING, FINISHED
    }
    protected ServerLevel level;
    protected BlockPos targetBlock;
    protected List<ServerPlayer> playersGroup;
    protected final UUID id = UUID.randomUUID();
    protected final Observer<RandomEventStatus> onChangeStatusObserver = new Observer<>();
    public RandomEventStatus status = RandomEventStatus.NOT_STARTED;
    protected int distanceThreshold = 650;
    protected int timeLimit = 20 * 60 * 10;
    protected int timePassed = 0;
    public List<RandomEventsCategory> categories = List.of(RandomEventsCategory.SPECIAL);
    public RandomEventsRarity rarity = RandomEventsRarity.COMMON;
    public String title = "Random Event";
    public String eventTag = "default_event";

    private final Consumer<TickEvent.PlayerTickEvent> playerTickConsumer = this::playerTick;

    public RandomEvent create() {
        return new RandomEvent();
    }

    public void initEvent(ServerLevel level, BlockPos targetBlock, List<ServerPlayer> playersGroup) {
        this.level = level;
        this.targetBlock = targetBlock;
        this.playersGroup = playersGroup;
        setEventStatus(RandomEventStatus.NOT_STARTED);
    }

    public void playerTick(TickEvent.PlayerTickEvent event) {
        if (Minecraft.getInstance().isPaused() || event.player.isSpectator() || !event.side.isServer()) return;

        onTimeLimit();
        if (isNearby(event.player) && status == RandomEventStatus.WAITING_PLAYER) {
            onPlayerEnter();
        }
    }

    public void start() {
        if (level == null || targetBlock == null) {
            throw new Error("Level and Target Block must be initialized");
        }
        onChangeStatusObserver.add(this::onChangeStatus);
        setEventStatus(RandomEventStatus.PREPARING);
        MinecraftEventsObservers.playerTickObserver.add(this.playerTickConsumer);
    }

    public void onTimeLimit() {
        timePassed++;
        if (timePassed >= timeLimit) {
            cancelEvent();
        }
    }

    public boolean isNearby(Player player) {
        double distance = player.blockPosition().distSqr(targetBlock);
        return distance <= distanceThreshold;
    }

    public void onChangeStatus(RandomEventStatus newStatus) {
        level.getServer().sendSystemMessage(Component.literal("Event " + eventTag + " status: " + newStatus));
        switch (newStatus) {
            case PREPARING -> onPrepare();
            case FINISHING_SUCCESS -> onFinishingSuccess();
            case FINISHING_CANCELED -> onFinishingCanceled();
            case FINISHING -> onFinishing();
            case READY -> onReady();
        }
    }

    public void setEventStatus(RandomEventStatus newStatus) {
        this.status = newStatus;
        onChangeStatusObserver.notify(status);
    }

    protected void onFinishingSuccess() {
        setEventStatus(RandomEventStatus.FINISHING);
    }

    protected void onFinishingCanceled() {
        MinecraftEventsObservers.playerTickObserver.remove(this.playerTickConsumer);
        setEventStatus(RandomEventStatus.FINISHING);
    }

    protected void onFinishing() {
        setEventStatus(RandomEventStatus.FINISHED);
    }

    protected void onPrepare() {
        setEventStatus(RandomEventStatus.WAITING_PLAYER);
    }

    public void onPlayerEnter() {
        MinecraftEventsObservers.playerTickObserver.remove(this.playerTickConsumer);
        setEventStatus(RandomEventStatus.READY);
    }

    protected void onReady() {}

    public void resolve() {
        setEventStatus(RandomEventStatus.FINISHING_SUCCESS);
    }

    public void cancelEvent() {
        this.setEventStatus(RandomEventStatus.FINISHING_CANCELED);
    }

    public boolean isEntityOfEvent(Entity entity) {
        CompoundTag nbt = entity.getPersistentData();
        return nbt.contains(eventTag) && nbt.getString(eventTag).equals(this.id.toString());
    }

    protected void addEntityToEvent(Entity entity) {
        CompoundTag nbt = entity.getPersistentData();
        nbt.putString(RandomEventsMod.ID, this.id.toString());
    }

    protected void removeEntityFromEvent(Entity entity) {
        CompoundTag nbt = entity.getPersistentData();
        nbt.remove(RandomEventsMod.ID);
    }
}

File: src/main/kotlin/endytkn/randomEvents/randomEvents/RandomEvent.kt
package endytkn.randomEvents.randomEvents

import endytkn.randomEvents.minecraftEventsObservers.MinecraftEventsObservers
import endytkn.randomEvents.titleManager.TitleManager
import endytkn.randomEvents.utils.GameInstances
import endytkn.randomEvents.utils.Observer
import net.minecraft.client.Minecraft
import net.minecraft.core.BlockPos
import net.minecraft.network.chat.Component
import net.minecraft.server.level.ServerLevel
import net.minecraft.server.level.ServerPlayer
import net.minecraft.world.entity.Entity
import net.minecraft.world.entity.LivingEntity
import net.minecraft.world.entity.player.Player
import net.minecraft.world.level.Level
import net.minecraftforge.event.TickEvent
import java.util.UUID

enum class RandomEventStatus {
     NOT_STARTED, PREPARING, READY, WAITING_PLAYER, FINISHING_SUCCESS, FINISHING_CANCELED, FINISHING, FINISHED
}

open class RandomEvent() {
    var level: ServerLevel? = null
    var targetBlock: BlockPos? = null
    var playersGroup: List<ServerPlayer>? = null
    val id: UUID = UUID.randomUUID()
    val onChangeStatusObserver = Observer<RandomEventStatus>()
    var status: RandomEventStatus = RandomEventStatus.NOT_STARTED
    var distanceThreshold = 50
    var timeLimit = 20 * 60 * 10
    var timePassed = 0
    var title: String = "Random Event"
    var category: RandomEventsCategories = RandomEventsCategories.SPECIAL
    var rarity: RandomEventsRarity = RandomEventsRarity.COMMON
    var eventTag = "randomEvent"

    open fun create(): RandomEvent {
        return RandomEvent()
    }

    open fun initEvent(level: ServerLevel, targetBlock: BlockPos, playersGroup: List<ServerPlayer>) {
        this.level = level
        this.targetBlock = targetBlock
        this.playersGroup = playersGroup
        setEventStatus(RandomEventStatus.NOT_STARTED)
    }

    open fun start() {
        if (level == null || targetBlock == null) throw Error("Level and Target Block Must be Initialized")

        onChangeStatusObserver += ::onChangeStatus
        setEventStatus(RandomEventStatus.PREPARING)
        MinecraftEventsObservers.onPlayerTickEventObserver.plusAssign(::tickEvent)
    }

    open fun tickEvent(event: TickEvent.PlayerTickEvent) {
        try {
            if (Minecraft.getInstance().isPaused || event.player.isSpectator) return
            onTimeLimit()
            if (isNearby(event.player) && event.side.isServer && status == RandomEventStatus.WAITING_PLAYER) {
                onPlayerEnter()
            }
        } catch (e: Error) {
            println(e)
            throw e;
        }
    }

    open fun onTimeLimit() {
        timePassed++;
        if (timePassed >= timeLimit) {
            cancelEvent()
        }
    }

    private fun isNearby(player: Player): Boolean {
        val distance = player.blockPosition().distSqr(targetBlock!!)
        return distance <= distanceThreshold
    }

    open fun onChangeStatus(newStatus: RandomEventStatus) {
        level!!.server!!.sendSystemMessage(Component.literal("Event $eventTag status: $newStatus"))
        when(newStatus) {
            RandomEventStatus.PREPARING -> { this.onPrepare() }
            RandomEventStatus.FINISHING_SUCCESS -> { this.onFinishingSuccess() }
            RandomEventStatus.FINISHING_CANCELED -> { this.onFinishingCanceled(); }
            RandomEventStatus.FINISHING -> { this.onFinishing(); }
            RandomEventStatus.READY -> { this.onReady() }
            else -> {}
        }

    }

    open fun setEventStatus(newStatus: RandomEventStatus) {
        this.status = newStatus
        onChangeStatusObserver.invoke(status)
    }

    open fun onFinishingSuccess() {
        setEventStatus(RandomEventStatus.FINISHING)
    }

    open fun onFinishingCanceled() {
        setEventStatus(RandomEventStatus.FINISHING)
    }

    open fun onFinishing() {
        setEventStatus(RandomEventStatus.FINISHED)
    }

    open fun onPrepare() {
        setEventStatus(RandomEventStatus.WAITING_PLAYER)
    }

    open fun onPlayerEnter() {
        MinecraftEventsObservers.onPlayerTickEventObserver.minusAssign(::tickEvent)
        setEventStatus(RandomEventStatus.READY)
    }

    open fun onReady() {}

    open fun resolve() {
        setEventStatus(RandomEventStatus.FINISHING_SUCCESS)
    }

    open fun cancelEvent() {
        MinecraftEventsObservers.onPlayerTickEventObserver.minusAssign(::tickEvent)
        setEventStatus(RandomEventStatus.FINISHING_CANCELED)
    }

    open fun isEntityOfEvent(entity: Entity): Boolean {
        val nbt = entity.persistentData
        return nbt.contains(this.eventTag) && nbt.getString(this.eventTag) == this.id.toString()
    }
}